### ISSUE TITLE
How does the package enforce CUDA version alignment?

### ISSUE THREAD

**Issue by danpetry** [conda-forge/cuda-version-feedstock#17]
### Comment:

AFAIU, this package a) wraps the __CUDA virtual package into something more friendly and b) enforces CUDA version alignment across an environment. Is that correct? How does it do b) for v12.x, when the cudatoolkit package doesn’t exist?

**Comment by bdice** [comment]
Hi @danpetry, good question. The `__cuda` virtual package ([documentation](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-virtual.html)) represents the "maximum version of CUDA supported by the display driver."

The `cuda-version` package controls what CUDA toolkit (and associated runtime libraries) are installed. It is meant to be a single control point for both end-users and package maintainers that forces all CUDA packages to come from the same major.minor version.

For CUDA 11, installing `cuda-version=11.8` or `cuda-version=11.5` constrains the installable versions of `cudatoolkit` to the same major.minor version.

For CUDA 12, the CUDA Toolkit has been split into separate packages. There are now packages like `libcublas`, `cuda-nvcc`, etc. These packages have a run dependency on a specific `cuda-version`. Note that components of the CUDA Toolkit aren't always versioned the same way as the CUDA Toolkit as a whole (not all packages shipped with CUDA 12.0 have the version number 12.0), so `cuda-version` provides a way to control the version alignment of the installed CUDA packages.

The issue here https://github.com/conda-forge/conda-forge.github.io/issues/1963 also documents some higher-level information about CUDA 12 on conda-forge and would be a good place to learn more about the structure of CUDA 12 on conda-forge.

There is a full list of components from CUDA 12.0 in this issue (https://github.com/conda-forge/staged-recipes/issues/21382) but be aware that things change slightly with new releases. New packages (or platform support for existing packages) may be added in CUDA 12.1, 12.2, and so on, to align with upstream CUDA Toolkit releases.

Note that CPU-only systems are permitted to install CUDA Toolkit packages. Currently the `cuda-version` package [has a `run_constrained` on `__cuda`](https://github.com/conda-forge/cuda-version-feedstock/blob/902045016fbc6e4dd1350a7390b0411f376d1a19/recipe/meta.yaml#L16) so that you can't install a CUDA Toolkit that is unsupported by your driver _if_ a CUDA driver is present.

In summary, the `__cuda` virtual package determines what versions of CUDA are supported by your host system's CUDA driver. That driver support determines what `cuda-version` and CUDA Toolkit packages you can install.

**Comment by danpetry** [comment]
Thanks both for the extensive information. This gives me plenty to go on.

**Comment by danpetry** [comment]
Can I ask why you used `run_constrained`, i.e. why you allow people to install cuda packages on a CPU-only machine?

**Comment by danpetry** [comment]
I suppose if a certain package wants to enforce this, then it can include `__CUDA` in the run section. Doing it like this gives downstream packages that choice?

**Comment by jakirkham** [comment]
In order to install packages during a build, we allow the non-CUDA case

**Comment by vyasr** [comment]
Closing as resolved.
